users_ldap_groups/models/res_company_ldap_operator.py

```diff
@@ -20,22 +20,33 @@ def operators(self):
         """Return names of function to call on this model as operator"""
         return ("contains", "equals", "query")
 
-    def contains(self, ldap_entry, mapping):
+    def _contains(self, ldap_entry, mapping):
         return mapping.ldap_attribute in ldap_entry[1] and mapping.value in map(
             lambda x: x.decode(), ldap_entry[1][mapping.ldap_attribute]
         )
 
-    def equals(self, ldap_entry, mapping):
+    def _equals(self, ldap_entry, mapping):
         return mapping.ldap_attribute in ldap_entry[1] and mapping.value == str(
             list(map(lambda x: x.decode(), ldap_entry[1][mapping.ldap_attribute]))
         )
 
-    def query(self, ldap_entry, mapping):
+    def _query(self, ldap_entry, mapping):
         query_string = Template(mapping.value).safe_substitute(
-            {attr: ldap_entry[1][attr][0].decode() for attr in ldap_entry[1]}
+            {
+                attr: self.safe_ldap_decode(ldap_entry[1][attr][0])
+                for attr in ldap_entry[1]
+            }
         )
 
         results = mapping.ldap_id._query(mapping.ldap_id.read()[0], query_string)
         _logger.debug('Performed LDAP query "%s" results: %s', query_string, results)
 
         return bool(results)
+
+    def safe_ldap_decode(self, attr):
+        import base64
+
+        try:
+            return attr.decode()
+        except UnicodeDecodeError:
+            return base64.b64encode(attr)
```

Comment:
A comment explaining why a base 64 decoding fallback makes sense would be helpful.

Reply:
A docstring explaining the decision made added for the method.

Reply:
The `x.decode()` is there in the `_contains()` & `_equals()`, too, and has been part of this module since it was migrated to 12.0 in [[MIG] users_ldap_groups: Migration to 12.0](https://github.com/OCA/server-auth/commit/ad15e3aa96d7d6f04aba840cd1499ad2c3af78ea) ([blame](https://github.com/OCA/server-auth/blame/4501d8a8772e90b99be07ff3a093f4036e6a1a72/users_ldap_groups/models/res_company_ldap_operator.py)). This PR never added it, but only the `base64.b64encode(attr)` as a fallback in case it fails.

Maybe @alexey-pelykh could shed some light on the background behind this decision.

Reply:
I concur with @oh2fih that such fix should be consistent across the codebase. First and foremost - do we have actual usecase when we need binary data and don't skip the non-UTF8-decodable string?

Reply:
> So, why not use this safe_ldap_decode() in contains & equals then? Because only the requested LDAP attributes are compared and they should never contain binary data. This is a special case for the query operator for which LDAP returns the entire Active Directory object with all the attributes.

Okay, that explains why only `_query` yet the part "do we actually have any need for binary fields like photos" remain

Reply:
No, we do not **need** binary fields, but we need to except to receive them, as I [explained](https://github.com/OCA/server-auth/pull/596#issuecomment-2147906595):

> This is a special case for the `query` operator for which LDAP returns the entire Active Directory object with all the attributes.

Reply:
I agree on debugging